File: playground/src/features/inspector/graph.ts
import type {
  InspectorFile,
  InspectorGraphEdge,
  InspectorGraphNode,
  InspectorGraphNodeKind,
} from "./types";

interface ImportRecord {
  specifier: string;
  kind: InspectorGraphEdge["kind"];
  locals: string[];
}

export function buildInspectorGraph(files: InspectorFile[]): {
  nodes: InspectorGraphNode[];
  edges: InspectorGraphEdge[];
} {
  const fileMap = new Map(files.map((file) => [normalizePath(file.path), file]));
  const nodes = files.map((file) => ({
    path: normalizePath(file.path),
    kind: fileKind(file.path),
    isEntry: isEntryPath(file.path),
    sourceLines: lineCount(file.source),
    sourceBytes: new TextEncoder().encode(file.source).length,
    issueCount: 0,
  }));
  const edges: InspectorGraphEdge[] = [];

  for (const file of files) {
    const from = normalizePath(file.path);
    for (const record of extractImports(file.source)) {
      const target = resolveImport(fileMap, from, record.specifier);
      if (!target) continue;

      pushEdge(edges, {
        from,
        to: target.path,
        kind: record.kind,
        specifier: record.specifier,
      });

      if (target.path.endsWith(".vue") && componentIsUsed(file.source, record.locals)) {
        pushEdge(edges, {
          from,
          to: target.path,
          kind: "component",
          specifier: record.specifier,
        });
      }
    }
  }

  return { nodes, edges };
}

function pushEdge(edges: InspectorGraphEdge[], edge: InspectorGraphEdge) {
  if (
    edges.some(
      (existing) =>
        existing.from === edge.from &&
        existing.to === edge.to &&
        existing.kind === edge.kind &&
        existing.specifier === edge.specifier,
    )
  ) {
    return;
  }
  edges.push(edge);
}

function extractImports(source: string): ImportRecord[] {
  const records: ImportRecord[] = [];
  const staticImportPattern = /import\s+([\s\S]*?)\s+from\s+["']([^"']+)["']/g;
  const dynamicImportPattern = /import\s*\(\s*["']([^"']+)["']\s*\)/g;
  let match: RegExpExecArray | null;

  while ((match = staticImportPattern.exec(source)) !== null) {
    records.push({
      specifier: match[2],
      kind: "import",
      locals: extractImportLocals(match[1]),
    });
  }

  while ((match = dynamicImportPattern.exec(source)) !== null) {
    records.push({
      specifier: match[1],
      kind: "dynamic-import",
      locals: [],
    });
  }

  return records;
}

function extractImportLocals(clause: string): string[] {
  const locals: string[] = [];
  const trimmed = clause.trim().replace(/^type\s+/, "");
  const defaultName = trimmed.split(",")[0]?.trim();
  if (defaultName && /^[A-Za-z_$][\w$]*$/.test(defaultName)) {
    locals.push(defaultName);
  }

  const namedMatch = trimmed.match(/\{([^}]+)\}/);
  if (namedMatch) {
    for (const part of namedMatch[1].split(",")) {
      const local = part
        .split(/\s+as\s+/)
        .pop()
        ?.trim();
      if (local && /^[A-Za-z_$][\w$]*$/.test(local)) {
        locals.push(local);
      }
    }
  }

  return locals;
}

function componentIsUsed(source: string, locals: string[]): boolean {
  return locals.some((local) => {
    const pascal = escapeRegExp(local);
    const kebab = escapeRegExp(toKebabCase(local));
    return new RegExp(`<\\s*(?:${pascal}|${kebab})(?:\\s|/|>)`).test(source);
  });
}

function resolveImport(
  fileMap: Map<string, InspectorFile>,
  from: string,
  specifier: string,
): { path: string; file: InspectorFile } | null {
  if (!specifier.startsWith(".")) return null;
  for (const candidate of importCandidates(from, specifier)) {
    const file = fileMap.get(candidate);
    if (file) return { path: candidate, file };
  }
  return null;
}

function importCandidates(from: string, specifier: string): string[] {
  const base = normalizePath(`${parentPath(from)}/${specifier}`);
  const candidates = [base];

  if (!hasKnownExtension(base)) {
    for (const extension of [".vue", ".ts", ".tsx", ".js", ".jsx"]) {
      candidates.push(`${base}${extension}`);
    }
    for (const extension of ["/index.vue", "/index.ts", "/index.js"]) {
      candidates.push(`${base}${extension}`);
    }
  }

  return candidates;
}

function normalizePath(path: string): string {
  const parts: string[] = [];
  for (const part of path.replace(/\\/g, "/").split("/")) {
    if (!part || part === ".") continue;
    if (part === "..") {
      parts.pop();
      continue;
    }
    parts.push(part);
  }
  return parts.join("/");
}

function parentPath(path: string): string {
  return path.includes("/") ? path.slice(0, path.lastIndexOf("/")) : "";
}

function fileKind(path: string): InspectorGraphNodeKind {
  if (path.endsWith(".vue")) return "vue";
  if (path.endsWith(".ts") || path.endsWith(".tsx")) return "typescript";
  if (path.endsWith(".js") || path.endsWith(".jsx")) return "javascript";
  return "other";
}

function isEntryPath(path: string): boolean {
  const basename = path.split("/").pop();
  return basename === "App.vue" || basename === "app.vue" || basename === "index.vue";
}

function hasKnownExtension(path: string): boolean {
  return /\.(vue|ts|tsx|js|jsx)$/.test(path);
}

function lineCount(source: string): number {
  return source ? source.split("\n").length : 0;
}

function toKebabCase(value: string): string {
  return value
    .replace(/([a-z0-9])([A-Z])/g, "$1-$2")
    .replace(/_/g, "-")
    .toLowerCase();
}

function escapeRegExp(value: string): string {
  return value.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
}

File: crates/vize_curator/src/inspector/graph.rs
//! Inspector module graph: nodes, edges, and path resolution.

use vize_s0::{String, ToCompactString};

use super::imports::analyze_file;
use super::payload::InspectorSourceFile;

#[derive(Debug, Clone, serde::Serialize)]
#[serde(rename_all = "camelCase")]
pub struct InspectorGraph {
    pub nodes: Vec<InspectorGraphNode>,
    pub edges: Vec<InspectorGraphEdge>,
}

#[derive(Debug, Clone, serde::Serialize)]
#[serde(rename_all = "camelCase")]
pub struct InspectorGraphNode {
    pub path: String,
    pub kind: &'static str,
    pub is_entry: bool,
    pub source_bytes: usize,
    pub source_lines: usize,
}

#[derive(Debug, Clone, PartialEq, Eq, serde::Serialize)]
#[serde(rename_all = "camelCase")]
pub struct InspectorGraphEdge {
    pub from: String,
    pub to: String,
    pub kind: &'static str,
    pub specifier: String,
}

pub fn build_graph(files: &[InspectorSourceFile]) -> InspectorGraph {
    let normalized_files: Vec<_> = files
        .iter()
        .map(|file| {
            let path = normalize_path(file.path.as_str());
            let analysis = analyze_file(path.as_str(), file.source.as_str());
            (path, file.source.as_str(), analysis)
        })
        .collect();

    let nodes = normalized_files
        .iter()
        .map(|(path, source, _)| InspectorGraphNode {
            path: path.clone(),
            kind: file_kind(path.as_str()),
            is_entry: is_entry_path(path.as_str()),
            source_bytes: source.len(),
            source_lines: line_count(source),
        })
        .collect();

    let mut edges = Vec::new();
    for (path, _, analysis) in &normalized_files {
        for import in &analysis.imports {
            if let Some(to) =
                resolve_import(&normalized_files, path.as_str(), import.specifier.as_str())
            {
                push_graph_edge(
                    &mut edges,
                    InspectorGraphEdge {
                        from: path.clone(),
                        to: to.clone(),
                        kind: import.kind,
                        specifier: import.specifier.clone(),
                    },
                );

                if import.kind == "import"
                    && is_component_module_path(to.as_str())
                    && component_is_used(&analysis.template_used_ids, &import.locals)
                {
                    push_graph_edge(
                        &mut edges,
                        InspectorGraphEdge {
                            from: path.clone(),
                            to,
                            kind: "component",
                            specifier: import.specifier.clone(),
                        },
                    );
                }
            }
        }
    }

    edges.sort_by(|left, right| {
        left.from
            .cmp(&right.from)
            .then_with(|| left.to.cmp(&right.to))
            .then_with(|| left.kind.cmp(right.kind))
            .then_with(|| left.specifier.cmp(&right.specifier))
    });

    InspectorGraph { nodes, edges }
}

fn component_is_used(template_used_ids: &vize_s0::FxHashSet<String>, locals: &[String]) -> bool {
    locals
        .iter()
        .any(|local| template_used_ids.contains(local.as_str()))
}

fn push_graph_edge(edges: &mut Vec<InspectorGraphEdge>, edge: InspectorGraphEdge) {
    if !edges.contains(&edge) {
        edges.push(edge);
    }
}

fn resolve_import(
    files: &[(String, &str, super::imports::FileAnalysis)],
    from: &str,
    specifier: &str,
) -> Option<String> {
    if !specifier.starts_with('.') {
        return None;
    }

    import_candidates(from, specifier)
        .into_iter()
        .find(|candidate| {
            files
                .iter()
                .any(|(path, _, _)| path.as_str() == candidate.as_str())
        })
}

fn import_candidates(from: &str, specifier: &str) -> Vec<String> {
    let base = normalize_path(join_path(parent_path(from).as_str(), specifier).as_str());
    let mut candidates = vec![base.clone()];

    if !has_known_extension(base.as_str()) {
        for extension in [
            ".vue", ".ts", ".tsx", ".mts", ".cts", ".js", ".jsx", ".mjs", ".cjs",
        ] {
            let mut candidate = base.clone();
            candidate.push_str(extension);
            candidates.push(candidate);
        }

        for extension in [
            "/index.vue",
            "/index.ts",
            "/index.tsx",
            "/index.mts",
            "/index.cts",
            "/index.js",
            "/index.jsx",
            "/index.mjs",
            "/index.cjs",
        ] {
            let mut candidate = base.clone();
            candidate.push_str(extension);
            candidates.push(candidate);
        }
    }

    candidates
}

fn parent_path(path: &str) -> String {
    path.rsplit_once('/')
        .map(|(parent, _)| parent.to_compact_string())
        .unwrap_or_default()
}

fn join_path(parent: &str, specifier: &str) -> String {
    if parent.is_empty() {
        specifier.to_compact_string()
    } else {
        let mut joined = parent.to_compact_string();
        joined.push('/');
        joined.push_str(specifier);
        joined
    }
}

fn normalize_path(path: &str) -> String {
    let mut parts = Vec::new();
    for part in path.split('/') {
        match part {
            "" | "." => {}
            ".." => {
                parts.pop();
            }
            _ => parts.push(part),
        }
    }
    parts.join("/").to_compact_string()
}

fn has_known_extension(path: &str) -> bool {
    path.rsplit_once('.').is_some_and(|(_, extension)| {
        matches!(
            extension,
            "vue" | "ts" | "tsx" | "mts" | "cts" | "js" | "jsx" | "mjs" | "cjs"
        )
    })
}

fn is_component_module_path(path: &str) -> bool {
    path.rsplit_once('.')
        .is_some_and(|(_, extension)| matches!(extension, "vue" | "tsx" | "jsx"))
}

fn file_kind(path: &str) -> &'static str {
    match path.rsplit_once('.').map(|(_, extension)| extension) {
        Some("vue") => "vue",
        Some("ts") | Some("tsx") | Some("mts") | Some("cts") => "typescript",
        Some("js") | Some("jsx") | Some("mjs") | Some("cjs") => "javascript",
        _ => "other",
    }
}

fn is_entry_path(path: &str) -> bool {
    matches!(
        path.rsplit('/').next().unwrap_or(path),
        "App.vue" | "app.vue" | "index.vue" | "main.ts" | "main.js"
    )
}

pub(super) fn line_count(source: &str) -> usize {
    if source.is_empty() {
        0
    } else {
        source.split('\n').count()
    }
}
